### PR TITLE
Implement and test tutorial event identifier

### PR DESCRIPTION
**Pull request by arthurshir**
## Changes
- Put reminder command variants into shared constants file
- Fix bug in tutorials constants file
- Implement tutorial event identifier
- Write unit tests for tutorial event identifier

## Tests
- tested tutorial event identifier

### CODE REVIEW

File: utils/constants.js [hunk 1]
```diff
@@ -5,11 +5,19 @@ const FRANCO_ARAB_REMINDME_VARIANTS = [
     'fakarnei', 'fakrnei', 'fkrnaei', 'fkrnei',
 ];
 
+const REMINDME_MIDDLEWARE_VARIANTS = ['r', 'remind me', ...FRANCO_ARAB_REMINDME_VARIANTS];
```
Comment (bubakazouba) [inline utils/constants.js:8]: this constant is not only used by the middleware

File: tutorial/event_identifier.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+const processTime = require("../nlp/processTime.js"),
+    logger = require("../logger.js"),
+    constants = require("./constants.js"),
+    shared_constants = require("../utils/constants.js"),
+    {parseTimezone} = require("../botfunctions/timezonecommand.js");
+
+const LAX_TZ = "America/Los_Angeles";
```
Comment (bubakazouba) [inline tutorial/event_identifier.js:7]: some comment explaining why ur hardcoding the timezone, i'm assuming the timezone doesn't matter here.

File: tst/tutorial/eventIdentifier.test.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        };
+        validate_utterance_to_expected_event_id_pairs(map);
+    });
+
+    it("should categorize non timezone/reminder utterances as `null`", () => {
+        let map = {
+            'hello world': null,
+            '/list': null,
+            'some more random text': null,
+            '/remindlksfjdslkfj welrkj': null,
```
Comment (bubakazouba) [inline tst/tutorial/eventIdentifier.test.js:40]: add 'race to the moon': null

that should fail

File: tutorial/event_identifier.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+const TIMEZONE_COMMAND_VARIANTS = ["timezone"];
+
+
+function isBackslashCommand(utterance, commandStrings) {
```
Comment (bubakazouba) [inline tutorial/event_identifier.js:11]: forward* slash

File: tutorial/event_identifier.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    } else {
+        return false;
+    }
+}
+
+function isTimezoneCommand(utterance) {
+    return isBackslashCommand(utterance, TIMEZONE_COMMAND_VARIANTS);
+}
+
+function isValidTimezoneCommand(utterance){
```
Comment (bubakazouba) [inline tutorial/event_identifier.js:23]: use `_` to mark private functions (same for other files)

File: tst/tutorial/eventIdentifier.test.js (hunk 3 above)
Comment (arthurshir) [reply tst/tutorial/eventIdentifier.test.js:40]: Hmm good catch, that evaluates to INVALID_REMINDER which is a bug

Comment (bubakazouba) [reply tst/tutorial/eventIdentifier.test.js:40]: add an edge case for empty string

File: tutorial/event_identifier.js (hunk 2 above)
Comment (arthurshir) [reply tutorial/event_identifier.js:7]: going to rename to "any_valid_tz"

File: tutorial/event_identifier.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    let timezoneInput = utterance.split(" ").slice(1).join(" "); // remove the first word ("/timezone")
```
Comment (bubakazouba) [inline tutorial/event_identifier.js:24]: can you think of a way to resolve the redundancy here? If this logic changes we have 2 places to change it

maybe something as simple as moving this inside `parseTimezone`

File: tutorial/event_identifier.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
+const processTime = require("../nlp/processTime.js"),
+    logger = require("../logger.js"),
+    constants = require("./constants.js"),
+    shared_constants = require("../utils/constants.js"),
+    {parseTimezone} = require("../botfunctions/timezonecommand.js");
+
+const LAX_TZ = "America/Los_Angeles";
+const TIMEZONE_COMMAND_VARIANTS = ["timezone"];
```
Comment (bubakazouba) [inline tutorial/event_identifier.js:8]: move to shared constants and have it used in timezonecommand.js

File: utils/constants.js (hunk 1 above)
Comment (arthurshir) [reply utils/constants.js:8]: Do you have any ideas for a better name for this? I was thinking something that indicates that this is used in the regex match... but had trouble coming up with a better name.

**Review by bubakazouba — CHANGES_REQUESTED** [review]

**Comment by arthurshir** [comment]
Add `_` prefix to "private" functions

File: tutorial/event_identifier.js (hunk 6 above)
Comment (arthurshir) [reply tutorial/event_identifier.js:24]: Yes, I'll do some refactoring to parseTimezone

File: tutorial/event_identifier.js (hunk 7 above)
Comment (arthurshir) [reply tutorial/event_identifier.js:8]: Yes good call. I was feeling hesitant about modifying the code of other files. But, since we have comprehensive testing I should overcome this feeling of hesitation.

File: README.md [hunk 8]
```diff
@@ -39,6 +39,17 @@
...
 
+### Dev Process for adding Bot Function Commands
+
+- Use the the helper functions within botfunctions/utils.js for:
+    - creation of the command finder regex.
+    - command recognition within utterances.
+    - stripping command inputs from an utterance.
+- Interface requirement: 
+    - export `COMMAND_REGEX` constant for other platform implementations (ex: fb messenger) to utilize.
+    - ...
```
Comment (bubakazouba) [inline README.md:50]: Remove the `...`, add a TODO if you meant to do this later

File: botfunctions/constants.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    'fakarney', 'fakrney', 'fkraney', 'fkrney',
+    'fakarni', 'fakrni', 'fkrani', 'fkrni',
+    'fakarnei', 'fakrnei', 'fkrnaei', 'fkrnei',
+];
+const ENGLISH_REMINDME_VARIANTS = [
+    'r', 'remind me', 'remindme', 'reminder'
+];
+const REMINDME_COMMAND_VARIANTS = [...ENGLISH_REMINDME_VARIANTS, ...FRANCO_ARAB_REMINDME_VARIANTS];
+
+const TIMEZONE_COMMAND_VARIANTS = ['timezone'];
```
Comment (bubakazouba) [inline botfunctions/constants.js:12]: add `settimezone`

File: botfunctions/timezonecommand.js [hunk 10]
```diff
@@ -24,6 +26,9 @@ Or by using the /timezone command:
 • <code>/timezone EST</code>`,
 };
 
+const COMMAND_REGEX = utils.makeRegexForCommandStrings(constants.TIMEZONE_COMMAND_VARIANTS);
+
+
```
Comment (bubakazouba) [inline botfunctions/timezonecommand.js:31]: Nit replace 2 extra lines with just 1 extra line

File: botfunctions/remindercommand.js [hunk 11]
```diff
@@ -90,33 +102,51 @@ function remindmeCallBack(ctx) {
...
     });
-
-    const COMMANDS_VARIANTS = ['r', 'remindme', 'remind', 'reminder', 
-        // also add franco arab ones that dont have spaces in them
-        ...(constants.FRANCO_ARAB_REMINDME_VARIANTS.filter(x => x.indexOf(" ") == -1)),
-    ];
-    for(let commandVariant of COMMANDS_VARIANTS) {
+    
+    for(let commandVariant of constants.REMINDME_COMMAND_VARIANTS) {
         bot.command(commandVariant, remindmeCallBack);
```
Comment (bubakazouba) [inline botfunctions/remindercommand.js:118]: is this going to be a problem for telegraf if we give it a command that contains spaces?
I'm assuming you ran the bot after this commit so just confirm that the bot ran successfully and processed any commands

File: botfunctions/timezonecommand.js [hunk 12]
```diff
@@ -157,13 +162,37 @@ function timezoneCommandCallback(ctx, language) {
...
+}
+
+
 module.exports = {
     addToBot: addToBot,
-    parseTimezone: parseTimezone
+    parseTimezone: parseTimezone,
+    COMMAND_REGEX: COMMAND_REGEX,
+    utteranceHasCommand: utteranceHasCommand,
+    stripCommandFromUtterance: stripCommandFromUtterance,
```
Comment (bubakazouba) [inline botfunctions/timezonecommand.js:195]: we shouldn't need to export stripCommandFromUtterance

File: tst/botfunctions/timezonecommand.test.js [hunk 13]
```diff
@@ -66,4 +75,89 @@ describe("parseTimezone", () => {
...
+            'timezone': true,
+            '   timezone Dubai': true,
+            '   timezone': true,
+            'TIMEZONE': true,
+            'timezonesdf': false,
+            'timezon': false,
+            'timezonesdf sdf': false,
+            '/timezonesdf': false,
+            '/timezon': false,
+            '/timezonesdf sdf': false,
```
Comment (bubakazouba) [inline tst/botfunctions/timezonecommand.test.js:99]: add just 1 more for /settimezone

File: botfunctions/timezonecommand.js (hunk 12 above)
Comment (arthurshir) [reply botfunctions/timezonecommand.js:195]: Shouldn't we use it in `msgr/handler.js`?

Comment (bubakazouba) [reply botfunctions/timezonecommand.js:195]: ah ur right, `parseTimezone` takes `commandInput` not `utterance`

File: botfunctions/remindercommand.js (hunk 11 above)
Comment (arthurshir) [reply botfunctions/remindercommand.js:118]: Tested, no problem.

note: adding `remind` to variants.
note: variants with spaces are not added as commands to the bot.

**Review by bubakazouba — APPROVED** [review]
